components/raftstore/src/store/peer_storage.rs

```diff
@@ -202,86 +193,74 @@ impl EntryCache {
-            + entries_mem_size;
     }
 
-    pub fn compact_to(&mut self, idx: u64) {
+    pub fn compact_to(&mut self, mut idx: u64) {
+        // Clean cached entries which have been already sent to apply threads. For example,
+        // if entries [1, 10), [10, 20), [20, 30) are sent to apply threads and `compact_to(15)`
+        // is called, only [20, 30) will still be kept in cache.
+        while let Some(cached_entries) = self.trace.pop_front() {
+            if cached_entries.range.start > idx {
+                self.trace.push_front(cached_entries);
+                let trace_len = self.trace.len();
+                let trace_cap = self.trace.capacity();
+                if trace_len < SHRINK_CACHE_CAPACITY && trace_cap > SHRINK_CACHE_CAPACITY {
+                    self.trace.shrink_to(SHRINK_CACHE_CAPACITY);
+                }
+                break;
+            }
+            *cached_entries.entries.lock().unwrap() = Default::default();
+            idx = cmp::max(cached_entries.range.end - 1, idx);
+        }
+
         let cache_first_idx = self.first_index().unwrap_or(u64::MAX);
         if cache_first_idx > idx {
             return;
         }
-        let mut drained_cache_entries_size = 0;
+
         let cache_last_idx = self.cache.back().unwrap().get_index();
-        // Use `cache_last_idx + 1` to make sure cache can be cleared completely
-        // if necessary.
+        // Use `cache_last_idx + 1` to make sure cache can be cleared completely if necessary.
         self.cache
-            .drain(..(cmp::min(cache_last_idx + 1, idx) - cache_first_idx) as usize)
-            .for_each(|e| {
-                drained_cache_entries_size +=
-                    (bytes_capacity(&e.data) + bytes_capacity(&e.context)) as i64
-            });
-        self.mem_size_change -= drained_cache_entries_size;
+            .drain(..(cmp::min(cache_last_idx + 1, idx) - cache_first_idx) as usize);
+
         if self.cache.len() < SHRINK_CACHE_CAPACITY && self.cache.capacity() > SHRINK_CACHE_CAPACITY
         {
-            let old_capacity = self.cache.capacity();
             // So the peer storage doesn't have much writes since the proposal of compaction,
             // we can consider this peer is going to be inactive.
             self.cache.shrink_to_fit();
-            self.mem_size_change += self
-                .get_cache_vec_mem_size_change(self.cache.capacity() as i64, old_capacity as i64)
         }
     }
 
-    fn update_mem_size_change_before_clear(&mut self) {
-        self.mem_size_change -= self
-            .cache
-            .iter()
-            .map(|e| (bytes_capacity(&e.data) + bytes_capacity(&e.context)) as i64)
-            .sum::<i64>();
-    }
-
-    fn get_cache_vec_mem_size_change(&self, new_capacity: i64, old_capacity: i64) -> i64 {
-        ENTRY_MEM_SIZE as i64 * (new_capacity - old_capacity)
-    }
-
-    fn get_total_mem_size(&self) -> i64 {
+    fn get_total_mem_size(&self) -> u64 {
```

Comment:
Do we need to consider the allocated memory of `trace`?

Reply:
+1, you comment `It's already counted when building` in Apply Msg, but I cannot get the place.